Cds

bmatthieu3 opened the pull request:
Hi, 

The module astroquery.cds aims to query the mocserver from cds. The mocserver (http://alasky.unistra.fr/MocServer/query) returns a list of datasets having data in a spatial region (can be a Polygon, a cone or a mocpy object describing a moc see https://github.com/cds-astro/mocpy ) and/or matching a given set of metadata constraints.

It is also possible to call specific services on the returned datasets like a simple cone search, tap service search. This feature uses pyvo to call the services on a specific data set and returns a astropy.table.Table object containing all the matched sources.

pep8speaks commented:
Hello @bmatthieu3! Thanks for submitting the PR.







 - In the file [`astroquery/cds/dataset.py`](https://github.com/astropy/astroquery/blob/4e64f5cebe873a30e6fa224de7fb0f46cf1c5cb9/astroquery/cds/dataset.py), following are the PEP8 issues :

> [Line 6:1](https://github.com/astropy/astroquery/blob/4e64f5cebe873a30e6fa224de7fb0f46cf1c5cb9/astroquery/cds/dataset.py#L6): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 57:5](https://github.com/astropy/astroquery/blob/4e64f5cebe873a30e6fa224de7fb0f46cf1c5cb9/astroquery/cds/dataset.py#L57): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 102:77](https://github.com/astropy/astroquery/blob/4e64f5cebe873a30e6fa224de7fb0f46cf1c5cb9/astroquery/cds/dataset.py#L102): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 116:1](https://github.com/astropy/astroquery/blob/4e64f5cebe873a30e6fa224de7fb0f46cf1c5cb9/astroquery/cds/dataset.py#L116): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file





 - In the file [`astroquery/cds/spatial_constraints.py`](https://github.com/astropy/astroquery/blob/4e64f5cebe873a30e6fa224de7fb0f46cf1c5cb9/astroquery/cds/spatial_constraints.py), following are the PEP8 issues :

> [Line 147:43](https://github.com/astropy/astroquery/blob/4e64f5cebe873a30e6fa224de7fb0f46cf1c5cb9/astroquery/cds/spatial_constraints.py#L147): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'

 - In the file [`astroquery/cds/tests/setup_package.py`](https://github.com/astropy/astroquery/blob/4e64f5cebe873a30e6fa224de7fb0f46cf1c5cb9/astroquery/cds/tests/setup_package.py), following are the PEP8 issues :

> [Line 3:1](https://github.com/astropy/astroquery/blob/4e64f5cebe873a30e6fa224de7fb0f46cf1c5cb9/astroquery/cds/tests/setup_package.py#L3): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

keflavich commented on astroquery/cds/core.py:
```diff
@@ -0,0 +1,201 @@
+        request_payload = dict()
+        if not isinstance(constraints, Constraints):
+            print("Invalid constraints. Must be of MOCServerConstraints type")
+            raise TypeError
```
best to just `raise TypeError("Invalid constraints. Must be of MOCServerConstraints type")`; we try to avoid print statements

keflavich commented on astroquery/cds/core.py:
```diff
@@ -0,0 +1,201 @@
+            return value
+
+    @staticmethod
+    def orderipix2uniq(n_order, n_pix):
```
could you add a docstring for this function?

keflavich commented on astroquery/cds/core.py:
```diff
@@ -0,0 +1,201 @@
+        if not verbose:
+            commons.suppress_vo_warnings()
+
+        r = response.json()
```
generally we prefer to avoid single-character variables.  Perhaps use `result` here instead?  or `json_data`?

keflavich commented on astroquery/cds/output_format.py:
```diff
@@ -0,0 +1,68 @@
+        elif format is OutputFormat.Type.record:
+            self.request_payload.update({'get': 'record'})
+
+        # parse fields
```
should this be indented, or is the indentation of the if statement incorrect?

bmatthieu3 commented:
Thank you for your fast answers.

@keflavich - I think query_region would have a lot of params because you can pass a cone (pos + radius), a polygon (str) or a moc (url or local path, or mocpy object) plus a string constraint for meta data constraint. Moreover there is the params related to the output of the query : (if the user wants the number of matching data sets, a list of their id, the full set of their meta data or just a subset)
You can see on the http interface of the CDS MOC server that there is a lot of params that can be tuneable : http://alasky.unistra.fr/MocServer/query
What do you think for me to do so that it can be well integrated in astroquery ? (maybe you know examples of now integrated services that had to deal with this problem too)

Thank you for your advices.

bsipocz commented:
@bmatthieu3 - The most simplest way is to require the region type, but then pass all the others optionally as kwargs. From the documentation point of view, it's not the most transparent (you can still list out the kwargs for all region type in the docstring but that can't really be double checked during testing), but basically the way we usually deal with the cases of these conditional parameters.

keflavich commented:
We like to have a simple `query_region` that accepts a cone (in practice, just a sky-projected circle) or a polygon plus any number of keyword arguments.  Most services support this approach in some form, and in my experience it is the best entry point for new users.  All the extra keywords can be documented in a more complete and general `query` tool, or in a `help` command.  This is essentially the same thing @bsipocz suggested.

bmatthieu3 commented:
I changed the interface of the cds.query_region method by adding an optional kwargs. You can find a first version of the generated doc here :
http://compassionate-jepsen-f7e5f1.bitballoon.com/cds/cds.html

Thanks again for your feedbacks.

bmatthieu3 commented:
Hi @bsipocz @keflavich ,

Did you have time to see the changes and the generated doc ? Thank you
Matthieu.

keflavich commented on docs/cds/cds.rst:
```diff
@@ -0,0 +1,59 @@
+
+* The number of the data sets.
+* A dictionary containing the lists of meta data for all the resulting data sets. These lists of meta data are each indexed by the ID of the data set they are referring to.
+* A MOC resulting from the union/intersection of all the MOCs of the resulting data sets.
```
have you defined the acronym MOC yet?  I don't see it.

keflavich replied:
Ah, it's defined below.  Best define it here, though: Multi-Order Coverage object?

keflavich commented on docs/cds/complex_query.rst:
```diff
@@ -0,0 +1,45 @@
+-  have the word 'CDS' in their IDs (constraint on the ``ID`` meta data)
+-  cover at most 1% of the sky (constraint on the ``moc_sky_fraction`` meta data)
+
+Meta data constraint is written like an algebraic expression involving the set of meta data to constraint. For more
```
constraint -> constrain

keflavich commented on docs/cds/cone_search_query.rst:
```diff
@@ -0,0 +1,206 @@
+
+- ``overlaps`` (default). The matching data sets are those overlapping the MOC region.
+- ``covers``. The matching data sets are those covering the MOC region.
+- ``enclosed``. The matching data sets are those enclosing the MOC region.
```
Should it be `encloses` since the others are `covers` and `overlaps`?  For consistency.

bmatthieu3 commented:
I am using pyvo to query some services on the datasets returned by the MOCServer (astroquery.cds.Dataset objects whose API is here http://friendly-mahavira-0b4e6a.bitballoon.com/api/astroquery.cds.dataset#astroquery.cds.Dataset ). Is this interesting to integrate this feature directly in astroquery or should it be done at a higher level (i.e. in a code importing astroquery.cds and pyvo) ? (btw I have seen Travis CI/AppVeyor need the pyvo dependency to continue).

Moreover I use mocpy https://github.com/cds-astro/mocpy : in case the user wants to query the MOCServer with a MOC or get the union of the MOCs from the matching datasets. This is also a dependency that CI tools may need to perform the tests later.

keflavich commented:
For the `.travis` tests, we might want a separate branch that installs these dependencies.  The tests should check whether pyvo and mocpy are installed, and they should be skipped if not.  We would like to have some tests that don't require pyvo or mocpy, though - is it possible to write some, perhaps using monkeypatching?  At least, basic tests of the parsing and query validation procedures?

Additional functionality to do more sophisticated things with the queries is welcome.  I'm not sure exactly what you were discussing, but it's OK to add more than just the barebones query tools here.

bmatthieu3 commented:
> For the .travis tests, we might want a separate branch that installs these dependencies.

Ok. Do I have to do something and if so do you know similar situations where dependencies had to be installed in the astroquery CI env so that I can see how to do ?

> The tests should check whether pyvo and mocpy are installed, and they should be skipped if not. We would like to have some tests that don't require pyvo or mocpy, though - is it possible to write some, perhaps using monkeypatching? At least, basic tests of the parsing and query validation procedures?
Thanks. In 38dc9d5, I skip the tests that use mocpy/pyvo (12 in total) using the following decorator
```
@pytest.mark.skipif('mocpy' not in sys.modules,
                    reason="requires the MOCPy library")
```
The tests seem to be well skipped as noted in https://travis-ci.org/astropy/astroquery/jobs/397402944 but I have two tests failing because it cannot find json files in cds/tests/data/ directory. Maybe it is the code in ``test_mocserver.py`` in charge of finding the path which does not work in the CI env (like the os.path) ? Do you have an idea ? thanks.
```
def data_path(filename):
    data_dir = os.path.join(os.path.dirname(__file__), 'data')
    return os.path.join(data_dir, filename)
```

> Additional functionality to do more sophisticated things with the queries is welcome. I'm not sure exactly what you were discussing, but it's OK to add more than just the barebones query tools here.
That was my question thanks : just was wondering if astroquery is just for querying or if we can add more features on the retrieved datasets from the MOCServer.

keflavich commented:
For the first, about travis: I said "branch" but I really meant a different entry in `.travis.yml` that installs the dependencies (mocpy, pyvo) and then runs the tests.  You can add those to the yml file in this PR; it should be reasonably clear how to do that?  Just add pyvo and mocpy to the PIP_DEPENDENCIES (or whatever it is) environmental variable for the particular travis entry.

Data not being found/included is likely a symptom of a missing or empty setup.py file or maybe __init__.py file in the tests/ directory; look at some of the others (e.g., vizier) to see how they do it.

cdeil commented:
Briefly talked about this face-to-face with @bmatthieu3 .

I think it would be better to take `region` object here and support [regions.PolygonSkyRegion](http://astropy-regions.readthedocs.io/en/latest/api/regions.PolygonSkyRegion.html) and [CircleSkyRegion](http://astropy-regions.readthedocs.io/en/latest/api/regions.CircleSkyRegion.html).
For the  `MOC`, I'm not sure yet, we'll discuss tomorrow. One idea could be to leave that part as-is for now in this PR.

Concerning the `SpatialConstrain` class, I think it can be deleted, it does almost nothing. The region to query payload dict translation can just happen in the method that takes the region object as input. If you keep it, I think it should be `SpatialConstraint` with a `t`.

Somehow the rebase created a ton of commits here. Suggest to close and open a new PR with only one or a few commits.

bsipocz commented:
> Somehow the rebase created a ton of commits here. Suggest to close and open a new PR with only one or a few commits.

It looks like that there was a merge commit, rather than a rebase. If it's helpful I can go ahead and try to rebase this here (but then @bmatthieu3's local branch need to be pulled from the rebased version).

bmatthieu3 commented:
> If it's helpful I can go ahead and try to rebase this here (but then @bmatthieu3's local branch need to be pulled from the rebased version).

Ok you can try this if you want, I can git stash my changes on my side before pulling.
Or I can reopen a new PR, it does not bother me.

bsipocz commented:
@bmatthieu3 - Rebase is done.

bmatthieu3 commented:
I updated the module. Changes are :

* All the code is in the core.py
* `query_region` takes either a [regions.CircleSkyRegion](https://astropy-regions.readthedocs.io/en/latest/api/regions.CircleSkyRegion.html#regions.CircleSkyRegion), a [regions.PolygonSkyRegion](https://astropy-regions.readthedocs.io/en/latest/api/regions.PolygonSkyRegion.html#regions.PolygonSkyRegion) or a [mocpy.MOC](https://mocpy.readthedocs.io/en/latest/api.html) object.
* I added a new method `find_data_sets` for retrieving data-sets based on the values of their meta-datas. This method does not take into account any regions but only expression on the meta-datas.
* CI tools pass (monkeypatching, assert on payloads in test_mocserver.py and tests querying the remote MOCServer in test_mocserver_remote.py).
* The doc (docs/cds/cds.rst) has been pushed and can be viewed at this link: http://gifted-northcutt-6f99ca.bitballoon.com/cds/cds.html 

@bsipocz @keflavich please let me know if there are things to fix before merging, thank you !

bsipocz commented on docs/cds/cds.rst:
```diff
@@ -0,0 +1,452 @@
+    ...                        meta_data="ID=*HST*")
+    >>> moc.plot(title="Union of the spatial coverage of all the Hubble surveys.")
+
+.. image:: ./HST_union.png
```
ideally this image could be generated on the fly based on the docs example. However, for that we need to make sure the doctests/build is run with remote data, so I suggest to do this in a follow-up PR, rather than holding up this one.

bsipocz commented on docs/cds/cds.rst:
```diff
@@ -0,0 +1,452 @@
+----------------------------------------------------
+The following packages are required for the use of this module:
+
+* `astropy-healpix`_
```
could you please also add these optional dependencies to the main list in `docs/index.rst`?

bsipocz commented on .travis.yml:
```diff
@@ -43,7 +43,7 @@ env:
         # pandas is used as a workaround of an astropy issue, sometimes it's required for the remote tests
         - CONDA_DEPENDENCIES_REMOTE='requests beautifulsoup4 html5lib keyring aplpy matplotlib pyregion jinja2 flask pandas'
-        - PIP_DEPENDENCIES='https://github.com/keflavich/httpbin/archive/master.zip pytest-dependency'
+        - PIP_DEPENDENCIES='https://github.com/keflavich/httpbin/archive/master.zip regions astropy_healpix mocpy pytest-dependency'
```
`regions` and `astropy-healpix` are conda packaged and are available on the astropy channel, so would prefer if they were to be added as conda dependencies above.

bsipocz replied:
I stay corrected, healpix is not yet available on the astropy channel, but I'm about to fix it.

bsipocz commented:
@bmatthieu3 - On, and one more thing, that is actually ideal for a follow-up PR is to switch on doctesting the documentation. Currently the infrastructure is not ideal for it, but it can be done, basically by adding `# doctest: +REMOTE_DATA` to the end of each and every example line that relies on remote data.
Given that this module is branch new, I suspect all the examples work without any issues, so hopefully it will be a quick and easy PR.

bmatthieu3 commented:
I added the dependencies to `index.rst` and moved regions from the PIP_DEP to CONDA_DEP and CONDA_DEP_REMOTE.
There is a travis failure for the PYTHON_VERSION=2.7 NUMPY_VERSION=1.9 KEYRING_VERSION='<12.0' when importing regions. Changing NP_VERSION to 1.10 makes the test pass.

bsipocz commented on docs/index.rst:
```diff
@@ -67,10 +67,13 @@ The following packages are required for astroquery installation & use:
 * `keyring <https://pypi.python.org/pypi/keyring>`_
 * `Beautiful Soup <https://www.crummy.com/software/BeautifulSoup/>`_
 * `html5lib <https://pypi.python.org/pypi/html5lib>`_
+* `regions <https://astropy-regions.readthedocs.io/en/latest/>`_
```
these are still optional dependencies for the package, so the it was better the way you added before this commit.

bmatthieu3 replied:
Yes, I will remove 9e0753e

bmatthieu3 commented on .travis.yml:
```diff
@@ -102,7 +102,7 @@ matrix:
         - os: linux
           stage: Tests with other Python/Numpy versions
-          env: PYTHON_VERSION=2.7 NUMPY_VERSION=1.9  KEYRING_VERSION='<12.0'
+          env: PYTHON_VERSION=2.7 NUMPY_VERSION=1.10  KEYRING_VERSION='<12.0'
```
If it can help the travis error was something like :
``ValueError: numpy.dtype has the wrong size, try recompiling.``
and occured when importing CircleSkyRegion, PolygonSkyRegion from regions in the core.py.

bmatthieu3 commented:
@bsipocz - I did the changes (i.e. dependencies as optional) and removed the commit where I changed the test PYTHON_VERSION=2.7 NUMPY_VERSION=1.9 KEYRING_VERSION='<12.0' to NUMPY_VERSION=1.10 so that you can see the full error in the travis logs. Hope it can help. Thanks.

bsipocz commented:
Thanks @bmatthieu3, this was most useful. The issue now is narrowed down to conda packaging as `regions` still support numpy 1.9, and pip installing indeed works. So I'm merging this now as is (pip installing worked on my fork), and patch the travis config later in master.
